aten/src/ATen/Context.h

```diff
@@ -250,11 +243,17 @@ static inline bool hasMKLDNN() {
 }
 
 static inline void manual_seed(uint64_t seed) {
-  globalContext().defaultGenerator(DeviceType::CPU).manualSeed(seed);
+  auto& gen = detail::getDefaultCPUGenerator();
+  // See Note [Thread-safety and Generators]
+  std::lock_guard<std::mutex> lock(gen->mutex_);
+  gen->set_current_seed(seed);
```

Comment:
You can reduce the scope of this lock by introducing a block scope. We should only take the lock out while we are setting the CPU seed, not while we're setting CUDA seed.